### ISSUE TITLE
What can we do to make on-the-fly syntax checking happen OOB?

### ISSUE THREAD

**Issue by alandmoore** [josteink/emacs-oob-reboot#36]
On-the-fly syntax checking is a pretty critical feature of any code editor.  It's easier to set up in Emacs than it used to be, but it would be nice if it was on by default for `prog-mode`s.

The two systems I know of currently are flymake and flycheck

- flycheck is much better, IMHO, but it's in MELPA so out of the running.  It can be pretty resource-demanding anyway.
- flymake is included in default emacs, I believe.  It doesn't seem very useful out of the box, though.

Of course both tools rely on third-party tools for the actual checking, if I understand rightly.  

Given all this, is there hope for at least passable OOB syntax checking in emacs for at least a few more popular languages?  Any thoughts?

**Comment by angrybacon** [comment]
As you said, linters rely on external software and for some (most?), there are multiple options. For that reason I don't think it should be on by default.

How does `flymake` behave by default when turned on? What does it do for which modes?

**Comment by alandmoore** [comment]
>How does flymake behave by default when turned on? What does it do for which modes?

I might be "holding it wrong", but it doesn't look like it does a whole lot of anything.  It is supposed to support C, C++, MS IDL, Java, Latex, HTML, XML, and Perl out of the box, but I loaded up a C file and typed no end of gibberish and bad syntax and got no response from flymake.

In the past I've had 3rd-party packages/code snippets to use flymake with Python and Javascript (calling external linters), but apparently it just isn't so great without some configuration.

I'm not really gunning for flymake here, mostly hoping someone will surprise me with an amazing 3rd option that we can enable.

**Comment by angrybacon** [comment]
From the Wiki:

> FlyMake performs on-the-fly syntax checks on the files being edited using the external syntax check tool (usually the compiler).

Perhaps Emacs couldn't find whatever (`cc`? `gcc`? My days of C are far) Emacs is supposed to compile C with, or you must set that first? I honestly have no clue.

Well, enabling linters could be as bad as showing trailing whitespaces, as in: it could just be noise to the user who just got an AIDS-sick file from his careless coworker. Yet again, it would promote prettier code.

**Comment by josteink** [comment]
> I'm not really gunning for flymake here, mostly hoping someone will surprise me with an amazing 3rd option that we can enable.

Indeed. I've long left flymake behind, because it just doesn't work that well. flycheck on the other hand works really well with lots of mode.

So yeah. Any third unknown but good option?

That said, having something which "just works" for C or C++ isn't going to happen anytime soon. For a flycheck-like thing to actually notice errors, and report them, it has to know how the thing is supposed to be compiled. Should it use plain gcc? Does it depend on other modules and libraries? Does it depend on cmake? GNU Autotools? Or some other build-system? Etc etc.

Basically, even *manually* settings up a universal linting-config  for C/C++ should be considered hard to impossible even for seasoned users. Getting it working out of the box is just unrealistic, so let's not spend too much effort on that.

**Comment by jackrusher** [comment]
@josteink should we try to get `flycheck` into ELPA?

**Comment by josteink** [comment]
That sounds like a big undertaking in its own, but if *someone* managed to get that done, they would have my full support in making it a package shipped and enabled by default :)

**Comment by alandmoore** [comment]
Before anyone gets their hopes up, https://github.com/flycheck/flycheck/issues/801

**Comment by justinjk007** [comment]
RIP

**Comment by leppert** [comment]
@alandmoore @justinjk007 Not sure if this is good news or worse news, but the aforementioned issue might not be a problem given that @lunaryorn has stepped down from maintaining the project (see: https://github.com/flycheck/flycheck/issues/1177). It's possible a new maintainer might be willing to tackle the issues outlined in https://github.com/flycheck/flycheck/issues/801.

**Comment by justinjk007** [comment]
@leppert good find.

**Comment by swsnr** [comment]
ELPA would require me to assign my copyright to the FSF which I won't do. Since my copyright covers most parts of Flycheck this excludes Flycheck from ELPA and by extension Emacs, short of a full rewrite, so I'm afraid Flycheck won't make it to ELPA.

I've got my reasons, and that's all I'll say.
